typedbSchemaBuilder/builder.py

```diff
@@ -0,0 +1,386 @@
+from collections import deque
+from .Type import Type
+from .Exceptions import SchemaChecker
+import copy
+
+class Builder:
+    def __init__(self) -> None:
+        self._schema = "define"
+        self._context = "?#"
+        self._query_log = deque()
+        self._query_id_generator = 1
+        self._types = {
+            "attribute": Type("attribute"),
+            "entity": Type("entity"),
+            "relation": Type("relation")
+        }
+        self._types["attribute"].add_super_type("thing")
+        self._types["attribute"].root_type="attribute"
+        self._types["entity"].add_super_type("thing")
+        self._types["entity"].root_type="entity"
+        self._types["relation"].add_super_type("thing")
+        self._types["relation"].root_type="relation"
+
+    def get_schema(self):        
+        if len(self._query_log) == 0:
+            raise Exception("Error: Schema is empty. Make changes to the schema before attempting GetSchema")
+        escaped_string = r"" + self._schema
+        decoded_string = bytes(escaped_string, "utf-8").decode("unicode_escape")
+        print(decoded_string)
+        checker = SchemaChecker (
+            schema=self._schema, query_log=self._query_log, types_=self._types
+        )
+        checker.test()
+        return decoded_string
+
+    def abstract(self, type_: str, qid: int=-1):
+        if self._context == type_:
+            if self._schema[-1] == ";":
+                self._schema = self._schema[:-1] + ","
+            self._schema += "\n    abstract;"
+        else:
+            self._context = type_
+            self._schema += "\n" + type_ + " abstract;"
+
+        query_log=copy.deepcopy(self._query_log)
```

Comment:
I believe we can find a better solution than copying the whole log every time.
And what's happening with the schema if the last command is incorrect?

Reply:
Throws exception and nothing gets appended to the self._query_log()